File: MIDTERM_Lab_2/uphsd-cs-undug-mohammadfaissal/linear-algebra/assignment-01/determinant_solver.js
/**
 * =====================================================
 * Student Name    : UNDUG, MOHAMMAD FAISSAL A.
 * Course          : Math 101 - Linear Algebra
 * Assignment      : Programming Assignment 1 - 3x3 Matrix Determinant Solver
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 18, 2026
 * GitHub Repo     : https://github.com/[your-username]/uphsd-cs-undug-mohammadfaissal
 * Runtime         : Node.js (run with: node determinant_solver.js)
 *
 * Description:
 *   JavaScript equivalent of DeterminantSolver.java. This script computes
 *   the determinant of the same hardcoded 3x3 matrix using cofactor expansion
 *   along the first row. All intermediate steps are logged to the console
 *   using console.log() for complete transparency of the solution process.
 * =====================================================
 */

// -- SECTION 1: Matrix Declaration -------------------------------------------
// The assigned 3x3 matrix is declared as a 2D JavaScript array.
// These values match the roster entry for UNDUG, MOHAMMAD FAISSAL A.
// Outer array = rows, inner arrays = individual row values.
const matrix = [
    [4, 2, 3],   // Row 1
    [5, 6, 1],   // Row 2
    [3, 4, 2]    // Row 3
];

// -- SECTION 2: Matrix Printer -----------------------------------------------
// Accepts a 3x3 array and prints it in a formatted table-like style.
// Uses template literals for clean string interpolation.
function printMatrix(m) {
    console.log("  +------------------+");
    m.forEach(row => {
        const fmt = row.map(v => v.toString().padStart(3)).join("  ");
        console.log(`  | ${fmt}   |`);
    });
    console.log("  +------------------+");
}

// -- SECTION 3: 2x2 Determinant Helper ---------------------------------------
// Computes the determinant of a 2x2 matrix given four scalar values.
// Called three times during the cofactor expansion step.
// Parameters: a, b = first row; c, d = second row of the 2x2 sub-matrix.
function computeMinor(a, b, c, d) {
    // 2x2 determinant formula: ad - bc
    return (a * d) - (b * c);
}

// -- SECTION 4: Step-by-Step Determinant Solver ------------------------------
// Main solving function. Accepts the 3x3 matrix and:
//   1. Prints the matrix clearly
//   2. Computes each 2x2 minor (M11, M12, M13)
//   3. Prints the computation of each minor with its arithmetic
//   4. Computes and prints each signed cofactor term
//   5. Sums the cofactors and prints the final determinant
//   6. Checks for a singular matrix (det = 0)
function solveDeterminant(m) {
    const line = "=".repeat(52);

    // Print problem header
    console.log(line);
    console.log("  3x3 MATRIX DETERMINANT SOLVER");
    console.log("  Student: UNDUG, MOHAMMAD FAISSAL A.");
    console.log("  Assigned Matrix:");
    console.log(line);
    printMatrix(m);
    console.log(line);
    console.log();
    console.log("  Expanding along Row 1 (cofactor expansion):");
    console.log();

    // -- Step 1: Minor M11 --
    // Delete row 0 and column 0 -> remaining elements: [1][1],[1][2],[2][1],[2][2]
    const minor11 = computeMinor(m[1][1], m[1][2], m[2][1], m[2][2]);
    console.log(
        `  Step 1 - Minor M11: det([${m[1][1]},${m[1][2]}],[${m[2][1]},${m[2][2]}])` +
        ` = (${m[1][1]}x${m[2][2]}) - (${m[1][2]}x${m[2][1]}) = ${m[1][1]*m[2][2]} - ${m[1][2]*m[2][1]} = ${minor11}`
    );

    // -- Step 2: Minor M12 --
    // Delete row 0 and column 1 -> remaining: [1][0],[1][2],[2][0],[2][2]
    const minor12 = computeMinor(m[1][0], m[1][2], m[2][0], m[2][2]);
    console.log(
        `  Step 2 - Minor M12: det([${m[1][0]},${m[1][2]}],[${m[2][0]},${m[2][2]}])` +
        ` = (${m[1][0]}x${m[2][2]}) - (${m[1][2]}x${m[2][0]}) = ${m[1][0]*m[2][2]} - ${m[1][2]*m[2][0]} = ${minor12}`
    );

    // -- Step 3: Minor M13 --
    // Delete row 0 and column 2 -> remaining: [1][0],[1][1],[2][0],[2][1]
    const minor13 = computeMinor(m[1][0], m[1][1], m[2][0], m[2][1]);
    console.log(
        `  Step 3 - Minor M13: det([${m[1][0]},${m[1][1]}],[${m[2][0]},${m[2][1]}])` +
        ` = (${m[1][0]}x${m[2][1]}) - (${m[1][1]}x${m[2][0]}) = ${m[1][0]*m[2][1]} - ${m[1][1]*m[2][0]} = ${minor13}`
    );

    // -- Cofactor Terms --
    // C11 gets +sign, C12 gets -sign, C13 gets +sign (alternating sign rule)
    const c11 =  m[0][0] * minor11;
    const c12 = -m[0][1] * minor12;
    const c13 =  m[0][2] * minor13;

    console.log();
    console.log(`  Cofactor C11 = (+1) x ${m[0][0]} x ${minor11} = ${c11}`);
    console.log(`  Cofactor C12 = (-1) x ${m[0][1]} x ${minor12} = ${c12}`);
    console.log(`  Cofactor C13 = (+1) x ${m[0][2]} x ${minor13} = ${c13}`);

    // -- Final Determinant --
    // Add all three cofactor terms together for the determinant
    const det = c11 + c12 + c13;
    console.log();
    console.log(`  det(M) = ${c11} + (${c12}) + ${c13}`);
    console.log(line);
    console.log(`  DETERMINANT = ${det}`);

    // -- Singular Matrix Check --
    // If det = 0, print a warning that the matrix cannot be inverted
    if (det === 0) {
        console.log("  The matrix is SINGULAR -- it has no inverse.");
    }
    console.log(line);
}

// -- SECTION 5: Program Entry Point ------------------------------------------
// Call the main solver function with the student's assigned matrix.
solveDeterminant(matrix);

File: MIDTERM_Lab_2/uphsd-cs-undug-mohammadfaissal/linear-algebra/assignment-01/DeterminantSolver.java
/**
 * =====================================================
 * Student Name    : UNDUG, MOHAMMAD FAISSAL A.
 * Course          : Math 101 - Linear Algebra
 * Assignment      : Programming Assignment 1 - 3x3 Matrix Determinant Solver
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 18, 2026
 * GitHub Repo     : https://github.com/[your-username]/uphsd-cs-undug-mohammadfaissal
 *
 * Description:
 *   This program computes the determinant of a hardcoded 3x3 matrix assigned
 *   to UNDUG, MOHAMMAD FAISSAL A. for Math 101. The solution is computed using
 *   cofactor expansion along the first row. Each intermediate step (2x2 minor,
 *   cofactor term, running sum) is printed to the console in a readable format.
 * =====================================================
 */
public class DeterminantSolver {

    // -- SECTION 1: Matrix Declaration -------------------------------------------
    // Declare the 3x3 matrix assigned to this student.
    // Values are hardcoded as a 2D integer array in row-major order.
    // These values come from the Section 4 student roster for UNDUG, MOHAMMAD FAISSAL A.
    static int[][] matrix = {
        { 4, 2, 3 },   // Row 1 of assigned matrix
        { 5, 6, 1 },   // Row 2 of assigned matrix
        { 3, 4, 2 }    // Row 3 of assigned matrix
    };

    // -- SECTION 2: 2x2 Determinant Helper ---------------------------------------
    // Computes the determinant of a 2x2 matrix given its four elements.
    // Formula: det = (a * d) - (b * c)
    // This method is called three times during cofactor expansion,
    // once for each of the three 2x2 minors of the first row.
    static int computeMinor(int a, int b, int c, int d) {
        // Apply the 2x2 determinant formula: ad - bc
        return (a * d) - (b * c);
    }

    // -- SECTION 3: Matrix Printer -----------------------------------------------
    // Prints the 3x3 matrix to the console in a formatted, readable layout.
    // Used at the start of output to display the problem clearly.
    static void printMatrix(int[][] m) {
        System.out.println("  +------------------+");
        for (int[] row : m) {
            System.out.printf("  |  %2d   %2d   %2d   |%n", row[0], row[1], row[2]);
        }
        System.out.println("  +------------------+");
    }

    // -- SECTION 4: Step-by-Step Determinant Solver ------------------------------
    // Computes the determinant of a 3x3 matrix using cofactor expansion
    // along the first row. Prints each step clearly:
    //   (a) Label the step (Minor M11, M12, M13)
    //   (b) Show the 2x2 sub-matrix
    //   (c) Show the arithmetic of the minor
    //   (d) Compute and display each signed cofactor term
    //   (e) Display the final sum and determinant value
    static void solveDeterminant(int[][] m) {

        // Print the header and the matrix
        System.out.println("=".repeat(52));
        System.out.println("  3x3 MATRIX DETERMINANT SOLVER");
        System.out.println("  Student: UNDUG, MOHAMMAD FAISSAL A.");
        System.out.println("  Assigned Matrix:");
        System.out.println("=".repeat(52));
        printMatrix(m);
        System.out.println("=".repeat(52));
        System.out.println();
        System.out.println("  Expanding along Row 1 (cofactor expansion):");
        System.out.println();

        // -- Step 1: Compute minor M11 --
        // Remove row 0 and column 0 to get the 2x2 minor.
        // The remaining elements are m[1][1], m[1][2], m[2][1], m[2][2].
        int minor11 = computeMinor(m[1][1], m[1][2], m[2][1], m[2][2]);
        System.out.printf("  Step 1 - Minor M11: det([%d,%d],[%d,%d]) = (%dx%d)-(%dx%d) = %d - %d = %d%n",
            m[1][1], m[1][2], m[2][1], m[2][2],
            m[1][1], m[2][2], m[1][2], m[2][1],
            m[1][1] * m[2][2], m[1][2] * m[2][1],
            minor11);

        // -- Step 2: Compute minor M12 --
        // Remove row 0 and column 1. Remaining: m[1][0], m[1][2], m[2][0], m[2][2].
        int minor12 = computeMinor(m[1][0], m[1][2], m[2][0], m[2][2]);
        System.out.printf("  Step 2 - Minor M12: det([%d,%d],[%d,%d]) = (%dx%d)-(%dx%d) = %d - %d = %d%n",
            m[1][0], m[1][2], m[2][0], m[2][2],
            m[1][0], m[2][2], m[1][2], m[2][0],
            m[1][0] * m[2][2], m[1][2] * m[2][0],
            minor12);

        // -- Step 3: Compute minor M13 --
        // Remove row 0 and column 2. Remaining: m[1][0], m[1][1], m[2][0], m[2][1].
        int minor13 = computeMinor(m[1][0], m[1][1], m[2][0], m[2][1]);
        System.out.printf("  Step 3 - Minor M13: det([%d,%d],[%d,%d]) = (%dx%d)-(%dx%d) = %d - %d = %d%n",
            m[1][0], m[1][1], m[2][0], m[2][1],
            m[1][0], m[2][1], m[1][1], m[2][0],
            m[1][0] * m[2][1], m[1][1] * m[2][0],
            minor13);

        // -- Cofactor Terms --
        // Apply the alternating sign pattern: C11 = +m[0][0] * minor11
        //                                     C12 = -m[0][1] * minor12
        //                                     C13 = +m[0][2] * minor13
        int c11 =  m[0][0] * minor11;
        int c12 = -m[0][1] * minor12;
        int c13 =  m[0][2] * minor13;

        System.out.println();
        System.out.printf("  Cofactor C11 = (+1) x %d x %2d = %d%n", m[0][0], minor11, c11);
        System.out.printf("  Cofactor C12 = (-1) x %d x %2d = %d%n", m[0][1], minor12, c12);
        System.out.printf("  Cofactor C13 = (+1) x %d x %2d = %d%n", m[0][2], minor13, c13);

        // -- Final Determinant --
        // Sum the three cofactor terms to get the determinant value.
        int det = c11 + c12 + c13;
        System.out.printf("%n  det(M) = %d + (%d) + %d%n", c11, c12, c13);
        System.out.println("=".repeat(52));
        System.out.printf("  DETERMINANT = %d%n", det);

        // -- Singular Matrix Check --
        // A determinant of zero means the matrix is singular (non-invertible).
        if (det == 0) {
            System.out.println("  The matrix is SINGULAR -- it has no inverse.");
        }
        System.out.println("=".repeat(52));
    }

    // -- SECTION 5: Entry Point --------------------------------------------------
    // The main method is the program's entry point.
    // It calls solveDeterminant() with the student's assigned matrix.
    public static void main(String[] args) {
        solveDeterminant(matrix);
    }

}
